ledger/acctonline.go

```diff
@@ -0,0 +1,846 @@
+	if err != nil {
+		return basics.MicroAlgos{}, err
+	}
+
+	onlineRoundParams := ao.onlineRoundParamsData[offset]
+	return basics.MicroAlgos{Raw: onlineRoundParams.OnlineSupply}, nil
+}
+
+// LookupOnlineAccountData returns the online account data for a given address at a given round.
+func (ao *onlineAccounts) LookupOnlineAccountData(rnd basics.Round, addr basics.Address) (data basics.OnlineAccountData, err error) {
+	oad, err := ao.lookupOnlineAccountData(rnd, addr)
+	if err != nil {
+		return
+	}
+
+	data.MicroAlgosWithRewards = oad.MicroAlgosWithRewards
+	data.VotingData.VoteID = oad.VotingData.VoteID
+	data.VotingData.SelectionID = oad.VotingData.SelectionID
+	data.VotingData.StateProofID = oad.VotingData.StateProofID
+	data.VotingData.VoteFirstValid = oad.VotingData.VoteFirstValid
+	data.VotingData.VoteLastValid = oad.VotingData.VoteLastValid
+	data.VotingData.VoteKeyDilution = oad.VotingData.VoteKeyDilution
+
+	return
+}
+
+// roundOffset calculates the offset of the given round compared to the current dbRound. Requires that the lock would be taken.
+func (ao *onlineAccounts) roundOffset(rnd basics.Round) (offset uint64, err error) {
+	if rnd < ao.cachedDBRoundOnline {
+		err = &RoundOffsetError{
+			round:   rnd,
+			dbRound: ao.cachedDBRoundOnline,
+		}
+		return
+	}
+
+	off := uint64(rnd - ao.cachedDBRoundOnline)
+	if off > uint64(len(ao.deltas)) {
+		err = fmt.Errorf("round %d too high: dbRound %d, deltas %d", rnd, ao.cachedDBRoundOnline, len(ao.deltas))
+		return
+	}
+
+	return off, nil
+}
+
+// roundParamsOffset calculates the offset of the given round compared to the onlineRoundParams cache. Requires that the lock would be taken.
+func (ao *onlineAccounts) roundParamsOffset(rnd basics.Round) (offset uint64, err error) {
+	// the invariant is that the last element of ao.onlineRoundParamsData is for round ao.latest()
+	startRound := ao.latest() + 1 - basics.Round(len(ao.onlineRoundParamsData))
+	if rnd < startRound {
+		err = &RoundOffsetError{
+			round:   rnd,
+			dbRound: startRound,
+		}
+		return
+	}
+
+	off := uint64(rnd - startRound)
+	if off >= uint64(len(ao.onlineRoundParamsData)) {
+		err = fmt.Errorf("round %d too high: dbRound %d, onlineRoundParamsData %d", rnd, startRound, len(ao.onlineRoundParamsData))
+		return
+	}
+
+	return off, nil
+}
+
+// lookupOnlineAccountData returns the online account data for a given address at a given round.
+func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.Address) (data ledgercore.OnlineAccountData, err error) {
+	ao.accountsMu.RLock()
+	needUnlock := true
+	defer func() {
+		if needUnlock {
+			ao.accountsMu.RUnlock()
+		}
+	}()
+	var offset uint64
+	var paramsOffset uint64
+	var rewardsProto config.ConsensusParams
+	var rewardsLevel uint64
+	var persistedData persistedOnlineAccountData
+
+	for {
+		currentDbRound := ao.cachedDBRoundOnline
+		currentDeltaLen := len(ao.deltas)
+		offset, err = ao.roundOffset(rnd)
+		inHistory := false
```

Comment:
nit: more idiomatic to check the err right after setting it
```suggestion
		inHistory := false
		offset, err = ao.roundOffset(rnd)
```